openquake/hazardlib/shakemap.py

```diff
@@ -0,0 +1,237 @@
+
+#  Copyright (c) 2018, GEM Foundation
+
+#  OpenQuake is free software: you can redistribute it and/or modify it
+#  under the terms of the GNU Affero General Public License as published
+#  by the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+
+#  OpenQuake is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU Affero General Public License for more details.
+
+#  You should have received a copy of the GNU Affero General Public License
+#  along with OpenQuake.  If not, see <http://www.gnu.org/licenses/>.
+from urllib.request import urlopen
+import math
+import numpy
+from scipy.stats import truncnorm
+from scipy import interpolate
+
+from openquake.hazardlib import geo, site, imt
+from openquake.hazardlib.shakemapconverter import get_shakemap_array
+
+F32 = numpy.float32
+SHAKEMAP_URL = 'http://shakemap.rm.ingv.it/shake/{}/download/{}.xml'
+PCTG = 100  # percent of g, the gravity acceleration
+
+
+def get_sitecol_shakemap(array_or_id, sitecol=None, assoc_dist=None):
+    """
+    :param array_or_id: shakemap ID or full shakemap array
+    :param sitecol: SiteCollection used to reduce the shakemap
+    :param assoc_dist: association distance
+    :returns: a pair (filtered site collection, filtered shakemap)
+    """
+    if isinstance(array_or_id, int):
+        with urlopen(SHAKEMAP_URL.format(array_or_id, 'grid')) as f1, \
+             urlopen(SHAKEMAP_URL.format(array_or_id, 'uncertainty')) as f2:
+            array = get_shakemap_array(f1, f2)
+    else:
+        array = array_or_id
+    if sitecol is None:  # extract the sites from the shakemap
+        return site.SiteCollection.from_shakemap(array), array
+
+    # associate the shakemap to the site collection
+    # TODO: forbid IDL crossing
+    bbox = (array['lon'].min(), array['lat'].min(),
+            array['lon'].max(), array['lat'].max())
+    sitecol = sitecol.within_bb(bbox)
+    data = geo.utils.GeographicObjects(array)
+    dic = data.assoc(sitecol, assoc_dist)
+    sids = sorted(dic)
+    return sitecol.filtered(sids), numpy.array([dic[sid] for sid in sids])
+
+
+# Here is the explanation of USGS for the units they are using:
+# PGA = peak ground acceleration (percent-g)
+# PSA03 = spectral acceleration at 0.3 s period, 5% damping (percent-g)
+# PSA10 = spectral acceleration at 1.0 s period, 5% damping (percent-g)
+# PSA30 = spectral acceleration at 3.0 s period, 5% damping (percent-g)
+# STDPGA = the standard deviation of PGA (natural log of percent-g)
+
+def spatial_length_scale(imt, vs30clustered):
+    """
+    :param imt: intensity measure type
+    :param vs30clustered: boolean
+    :returns: the `b` parameter in the correlation matrix
+    """
+    T = imt.period
+    if T < 1:
+        b = 40.7 - 15.0 * T if vs30clustered else 8.5 + 17.2 * T
```

Comment:
Please add a reference to a scientific paper that explains those numbers

Reply:
Vitor says that the article is still in the publishing process, accepted but who knows when it will appear.

Reply:
Isn't this the same model that is already implemented here https://github.com/gem/oq-engine/blob/master/openquake/hazardlib/correlation.py#L126 ? If that is the case the scientific paper is Jayaram & Baker (2009) as cited in the doc string of that class.

Any way to get around having two different implementations of the same formula?